lib/workflow/stores/mongo.js

```diff
@@ -677,21 +693,91 @@ function mongoStoreFactory(waterline, Promise, Constants, Errors, assert, _) {
+                    $in:_.difference(Constants.Task.FinishedStates, [data.state])
+                };
+                
+                return waterline.taskdependencies.find(query).then(function(res){
+                    if (!_.isEmpty(res)){
+                        // Mongo update, remove task under dependent.anyOf
+                        var updateRemove = {
+                            $unset:{}
+                        };
+                        updateRemove.$unset['dependencies.anyOf.'+data.taskId] = "";
+                        var queryRemove = {
+                            graphId: data.graphId,
+                            reachable: true,
+                            // List of task Ids that need to remove dependent.anyOf.<data.taskId>
+                            $or:[]
+                        };
+
+                        // Mongo update, mark the downstream task unreachable
+                        var setUnreachable = {
+                            $set: {
+                                reachable: false
+                            }
+                        };
+                        var querySetUnreachable = {
+                            graphId: data.graphId,
+                            reachable: true,
+                            // List of task Ids that need to mark as unrechable
+                            $or:[]
+                        };
+                        
+                        // res is an array of tasks that depend on data.taskId
+                        _.forEach(res, function(depItr){
+                            var depNum = Object.keys(depItr.dependencies.anyOf).length;
+                            if (depNum > 1) {
+                                var taskToRemoveDep = {};
+                                taskToRemoveDep.taskId = depItr.taskId;
+                                queryRemove.$or.push(taskToRemoveDep);
+                            }
+                            else {
+                                var taskToSetUnreachable = {};
+                                taskToSetUnreachable.taskId = depItr.taskId;
+                                querySetUnreachable.$or.push(taskToSetUnreachable);
+                            }
+                        });
+                        
+                        var retPromise = [];
+                        if (queryRemove.$or.length > 0) {
+                            retPromise.push(waterline.taskdependencies
+                            .updateMongo(queryRemove, updateRemove, options));
+                        }
+                        if (querySetUnreachable.$or.length > 0) {
+                            retPromise.push(waterline.taskdependencies
+                            .updateMongo(querySetUnreachable, setUnreachable, options));
+                        }
+
+                        return Promise.all(retPromise);
+                    }  
+                });
+            }
+            else {
```

Comment:
The same problem with `updateDependentTasks`, the same task id may exist in two places, both places need be handled.